### PR TITLE
[BACKLOG-14077] 

### PR DESCRIPTION
**Pull request by diogofscmariano**
@afrjorge @dcleao can you please review the js part?

### CODE REVIEW

File: core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js [hunk 1]
```diff
@@ -77,21 +97,109 @@ define([
...
+        if (cd.stacked == true || fullName == "area") {
+          fullName += "Stacked";
+        }
+      }
+
+      return fullName;
+    },
+
     render: function(values) {
+      var myself = this;
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js]: On another end, I would not remove the base class' `renderChart` method: there are dashboards out there overriding this (not very orthodox, I know) and I think there might be other ways to handle this.

Here's an alternative that I think would work better:

```js
var chart = {
  renderChart: function() {		
    
    // Always executes asynchronously.

    var cd = this.chartDefinition;		
    
    if(this.dashboard.isValidQueryDefinition(cd)) {
      
      this.beginQuery(cd, _.bind(this._renderOuter, this));

    } else if(this.valuesArray != null) {

      this.execute(_.bind(function() { this._renderOuter(this.valuesArray); }, this));

    } else {
      // Initialize the component only
      this.execute(_.bind(this._renderOuter, this));
    }
  },

  _shouldUseVizApiStyles: function() {
    return !!this._vizApiStyles && !_.contains(chartTypesBlackList, this.getCccChartName());
  },

  _renderOuter: function(data) {

    var promiseExtensions;
       
    this._preProcessChartDefinition();

    if(this._shouldUseVizApiStyles()) {

      promiseExtensions = context.getAsync("pentaho/ccc/visual/" + this.getCccChartName())
          .then(function(View) {
             return View.type.extensionEffective;
          });
    } else {
      // no external extensions
      promiseExtensions = Promise.resolve(null);
    }

    promiseExtensions
      .then(_.bind(this._renderInner, this, data))
      .then(_.bind(this.endExec, this), _.bind(this.failExec, this));
  },

  render: function(data) {
    this._renderInner(data, null);
  },

  // Similar to yours `_internalRender` and to the old "render" method.
  _renderInner: function(data, externalChartDefinition) {
    
    $("#" + myself.htmlObject).append('<div id="' + myself.htmlObject + 'protovis"></div>');

    // Always clone the original chartDefinition.
    var cd = $.extend({}, this.chartDefinition);

    if(externalChartDefinition) cd = $.extend(externalChartDefinition, cd);

    cd.canvas = this.htmlObject + "protovis";

    // Process extension points
    if(cd.extensionPoints) {
      var ep = {};
      cd.extensionPoints.forEach(function(a) { ep[a[0]] = a[1]; });
      cd.extensionPoints = ep;
    }

    if(this._shouldUseVizApiStyles() && cd.colors == null) {
      cd.colors = paletteRegistry.get("cdf_default").colors;
    }

    this.chart = new this.cccType(cd);

    if(arguments.length > 0) {
      this.chart.setData(data, {crosstabMode: this.crosstabMode, seriesInRows: this.seriesInRows});
    }

    this.chart.render();
  }
}
```

File: core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+            format: {
+              number: {
+                style: numberStyle
+              }
+            },
+
+            errorMessage_visible: false,
+            noDataMessage_visible: false,
+            invalidDataMessage_visible: false,
+
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js]: These are probably not good defaults for CDF

File: core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js [hunk 3]
```diff
@@ -77,21 +97,109 @@ define([
...
+        myself._internalRender.apply(myself, [values, View.type.extensionEffective]);
+      }).then(_.bind(myself.endExec, myself), _.bind(myself.failExec, myself));
+    },
 
-      var o = $.extend({}, this.chartDefinition);
+    /**
+     *
+     */
+    _internalRender: function(values, externalChartDefinition){
+      var o = externalChartDefinition ? $.extend(externalChartDefinition, this.chartDefinition) : this.chartDefinition;
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js]: Still not sure that simply overwriting whatever the user might have specified in `preProcess` and `postFetch` is the way to go.

File: core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js [hunk 4]
```diff
@@ -0,0 +1,646 @@
...
+            plotFrameVisible: false,
+            plotSizeMin: {width: 600, height: 300},
+
+            clickable:  false,
+            selectable: false,
+            hoverable:  true,
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js]: `clickable` should be removed.
Not sure if `selectable` should be set to `false` — only if some of the styling changes we did don't work well with selection...

**Comment by diogofscmariano** [comment]
@dcleao can you please recheck? Still need to add unit tests and doc

File: core-js/src/main/javascript/cdf/components/ProtovisComponent.js [hunk 5]
```diff
@@ -23,11 +23,11 @@ define([
       if(this.parameters == undefined) {
         this.parameters = [];
       }
-
+      
       this.renderChart();
     },
 
-    render: function(values) {
+    _renderOuter: function(values) {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ProtovisComponent.js]: If you see my comment below, maybe we can just call `render` to `_renderOuter` and not need both. So this rename here would not be needed.

File: core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js [hunk 6]
```diff
@@ -14,52 +14,147 @@
...
+  var _stackedExceptions = ['barNormalized'];
 
   return ChartComponent.extend({
 
     query: null,
     chart: null,
 
-    _preProcessChartDefinition: function() {
+    _cccChartName: null,
+    getCccChartName: function () {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js]: Still think this name is not appropriate. Did you see my suggestion in the previous review?

File: core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js [hunk 7]
```diff
@@ -14,52 +14,147 @@
...
+    /**
+     *
+     * @param name
+     * @returns {*}
+     * @private
+     */
+    _getCccFullName: function (name) {
+      var fullName;
+      //get the ccc name first
+      var m = name.split(_reCccName);
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js]: Use of split to do this doesn't seem right — not the right tool for the job, even if you can get it to do what you want. For this use case, split returns an array of 4 entries...

File: core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js [hunk 8]
```diff
@@ -68,49 +163,81 @@ define([
...
-        pv.listenForPageLoad(function() {
+        pv.listenForPageLoad(function () {
           me.renderChart();
         });
       }
     },
 
-    render: function(values) {
+    render: function (data) {
+      this._renderOuter(data);
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js]: I might have complicated thing a bit, while trying to keep `render` backwards compatible... I think we can just live with renaming the current `_renderOuter` to the old name `render`. No? The difference will be semantic. While, before, `render` was itself synchronous and did not finish the execution, now, it is asynchronous and will finish the execution...

File: core-js/src/main/javascript/cdf/components/ChartComponent.js [hunk 9]
```diff
@@ -84,13 +84,15 @@ define([
     renderChart: function() {
       var cd = this.chartDefinition;
 
-      if(this.dashboard.isValidQueryDefinition(cd)) {
-        this.triggerQuery(cd, _.bind(this.render, this));
-      } else if(this.valuesArray != undefined) {
-        this.synchronous(_.bind(function() { this.render(this.valuesArray); }, this));
+      if (this.dashboard.isValidQueryDefinition(cd)) {
+        this.beginQuery(cd, _.bind(this._renderOuter, this));
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ChartComponent.js]: Keep calling the `render` method here (see comments below).

File: core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js [hunk 10]
```diff
@@ -68,49 +163,81 @@ define([
...
+    
+    _renderInner: function (data, externalChartDefinition) {
       $("#" + this.htmlObject).append('<div id="' + this.htmlObject + 'protovis"></div>');
 
-      this._preProcessChartDefinition();
+      // Always clone the original chartDefinition.
+      var cd = $.extend({}, this.chartDefinition);
+
+      if (externalChartDefinition) {
+        cd = $.extend(externalChartDefinition, cd);
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js]: I know I wrote this bit... but it is intended that chartDefinition prevails right? `externalChartDefinition` are default values. Just checking.

File: core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js [hunk 11]
```diff
@@ -0,0 +1,638 @@
...
+
+            // . dot
+            dot_lineWidth: 0,
+            dot_fillStyle: function() {
+              var c = this.delegate();
+              return this.finished(c && !this.scene.isActive ? c.alpha(0.5) : c);
+            }
+          },
+
+          props: {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js]: `props` are of the `model`, not the `view`. Please remove.

File: core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js [hunk 12]
```diff
@@ -0,0 +1,638 @@
...
+            bar_fillStyle: function() {
+              var c = this.delegate();
+              return this.finished(c && this.scene.isActive ? c.alpha(0.5) : c);
+            }
+          }
+        }
+      },
+
+      // [Visual Roles] - Bars, Columns, Line, Area
+      {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js]: `props` are of the `model`, not the `view`. The whole rule can be removed.

File: core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js [hunk 13]
```diff
@@ -0,0 +1,638 @@
...
+
+      // Heat Grid
+      {
+        select: {
+          priority: -1,
+          application: "pentaho-cdf",
+          type: "pentaho/ccc/visual/heatGrid"
+        },
+        apply: {
+          props: {
```
Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ccc/vizApi.conf.js]: `props` are of the `model`, not the `view`. Please remove.

File: core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js (hunk 10 above)
Comment (diogofscmariano) [inline core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js]: Yes. The objective is to have chartDefinition prevailing. The idea is that whatever the dev does with its component (and we at compat 3 mode), the moment we get into the render the viz api definitions will be smashed into the chart definitions, not overriding what was there. In parallel with this, the proposal is to remove some ccc property defaults to allow that when the devs dont change their property (either in CDE or CDF) the Viz Api definitions come in. Does this makes sense?

File: core-js/src/main/javascript/cdf/components/ChartComponent.js [hunk 14]
```diff
@@ -18,32 +18,34 @@ define([
...
-              if($.isArray(value) && value.length == 1 && ('' + value[0]).indexOf(';') >= 0) {
+              if ($.isArray(value) && value.length == 1 && ('' + value[0]).indexOf(';') >= 0) {
                 // Special case where single element will wrongly be treated as a parsable array by cda
-                value = Utils.doCsvQuoting(value[0],';');
+                value = Utils.doCsvQuoting(value[0], ';');
               }
               //else Will not be correctly handled for functions that return arrays
 
-              if(typeof value == 'function') { value = value(); }
+              if (typeof value == 'function') {
```
Comment (afrjorge) [inline core-js/src/main/javascript/cdf/components/ChartComponent.js]: please change to `===` or `_.isUndefined(value)`

Comment (dcleao) [inline core-js/src/main/javascript/cdf/components/ChartComponent.js]: I vote on `===`.

File: pentaho-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.ext.js [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+   * @private
+   */
+  var _reCccName = /^(.*)Chart$/;
+
+  var _context = PentahoTypeContext.getInstance();
+
+  /**
+   * List of viz types to exclude from the Viz Api chart defaults
+   *
+   * @type {[String]}
```
Comment (dcleao) [inline pentaho-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.ext.js]: `{string[]}` or `Array.<string>`

File: pentaho-js/src/test/javascript/cdf/components/ccc/BaseCccComponent.ext-spec.js [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+ */
+
+define([
+  'cdf/components/ccc/BaseCccComponent.ext'
+], function (BaseCccComponentExt) {
+
+  /**
+   * ## The Base Ccc Component Ext
+   */
+  fdescribe("The Base Ccc Component Ext #", function () {
```
Comment (dcleao) [inline pentaho-js/src/test/javascript/cdf/components/ccc/BaseCccComponent.ext-spec.js]: forgotten `fdescribe`.

**Comment by diogofscmariano** [comment]
@afrjorge @dcleao pr updated
i will check the cgg case in another pr. But it should apply the viz api styles

**Pull request merged**